File: app/src/main/java/com/doxigo/muchtoman/Links.kt
package com.doxigo.muchtoman

import androidx.room.ColumnInfo
import androidx.room.Dao
import androidx.room.Entity
import androidx.room.Index
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.PrimaryKey
import androidx.room.Query

/**
 * Transfers and duplicates — the two ways one movement of money shows up as more than one
 * transaction, and the two ways a total goes quietly wrong if nothing notices.
 *
 * Every detector here is a pure function of the transactions, and nothing guesses: a link is
 * applied to the totals only when nothing about it is ambiguous — [LinkCandidate.auto] — and her
 * verdict beats everything in both directions. A near miss is stored as a non-auto candidate,
 * and that is all it does: it counts in nothing, [LinkCandidateDao.touching] reads it beside a
 * row, and a verdict of hers has a pair to land on so the same question cannot come back.
 * Refund and fee detectors lived here once, computing links nothing consumed; a detector whose
 * output feeds no total and no screen is not caution, it is dead weight dressed as caution.
 */

object LinkKind {
    const val TRANSFER = "transfer"
    const val DUPLICATE = "duplicate"
}

object Verdict {
    const val CONFIRMED = "confirmed"
    const val REJECTED = "rejected"
}

/** A pair the detectors propose. Rebuilt on every derive, so it lives in `derived.db`. */
@Entity(
    tableName = "link_candidate",
    primaryKeys = ["a_ref", "b_ref", "kind"],
    indices = [Index("a_ref"), Index("b_ref")],
)
data class LinkCandidate(
    @ColumnInfo(name = "a_ref") val aRef: String,
    @ColumnInfo(name = "b_ref") val bRef: String,
    val kind: String,
    /** Why, in a word, so the deck can say what it noticed rather than just assert it. */
    val reason: String,
    /** True only when nothing about the match is ambiguous. False means: not certain enough to apply. */
    val auto: Boolean,
    /**
     * The leg that arrived second, which is the one [hiddenRefs] strikes out. [aRef]/[bRef]
     * order by ref, and a ref is a hash — its order against time is a coin flip — so time is
     * carried here rather than inferred. Blank only on a pair built from a verdict whose legs
     * the ledger no longer holds.
     */
    @ColumnInfo(name = "later_ref", defaultValue = "''") val laterRef: String = "",
)

/**
 * Her verdict on a pair, which always wins.
 *
 * [aRef] is always lexicographically less than [bRef]. That ordering is load-bearing rather
 * than tidy: without it `(a,b)` and `(b,a)` are two different rows and the same pair can be
 * confirmed and rejected at the same time.
 */
@Entity(
    tableName = "link_decision",
    indices = [Index("a_ref", "b_ref", "kind", unique = true), Index("updated_at")],
)
data class LinkDecision(
    @PrimaryKey val id: String,
    @ColumnInfo(name = "a_ref") val aRef: String,
    @ColumnInfo(name = "b_ref") val bRef: String,
    val kind: String,
    val verdict: String,
    @ColumnInfo(name = "created_at") val createdAt: Long,
    @ColumnInfo(name = "updated_at") val updatedAt: Long,
    val deleted: Boolean = false,
)

fun linkDecision(aRef: String, bRef: String, kind: String, verdict: String, now: Long): LinkDecision {
    val (a, b) = if (aRef <= bRef) aRef to bRef else bRef to aRef
    return LinkDecision(uuid7(now), a, b, kind, verdict, now, now)
}

/** A fee band that a bank may take out of a transfer without it stopping being the same money. */
const val TRANSFER_FEE_MAX_RIAL = 120_000L

/** Instant rails settle in seconds; anything past this is two separate movements. */
const val TRANSFER_WINDOW_MS = 15 * 60 * 1000L

/**
 * پایا settles in batches, genuinely next-business-day, and ساتنا in banking-hour cycles.
 * Neither is the seconds that [TRANSFER_WINDOW_MS] assumes. Too wide to ever trust silently —
 * a pair found this way is always a question for the deck, never an answer.
 */
const val SLOW_RAIL_WINDOW_MS = 40 * 60 * 60 * 1000L

/**
 * The rails that do not settle while she watches.
 *
 * Read off *either* leg, because the two banks describe one transfer differently: the receiving
 * bank is usually the one that writes «واریز پایا», while the sending bank says only «انتقال».
 * Taking the rail from the sender alone gave those transfers the fifteen-minute instant window,
 * and a leg that landed six hours later was never paired with anything.
 */
private val SLOW_RAILS = setOf("paya", "satna")

/** Two identical purchases this close together are a real thing; only she may decide. */
const val DUPLICATE_WINDOW_MS = 90 * 1000L

/**
 * A reference number is the bank's own idempotency key, but some banks re-print a contract or
 * instalment number where one belongs, so even this match carries a window. Seven days: wide
 * enough for a delayed second copy of a real message, far short of next month's instalment.
 */
const val DUPLICATE_REFNO_WINDOW_MS = 7L * 24 * 60 * 60 * 1000

/**
 * Two legs sharing a post-transaction balance settle a duplicate only this close together.
 *
 * Unbounded, this match also caught a backup-restore that re-stamps DATE months later — and that
 * protection is deliberately traded away, because a fixed rent cycling an account back to the
 * same figure every month is normal use, and it was being auto-hidden from every report for
 * ever. A re-stamped restore now shows as two rows months apart, which is at least a wrongness
 * she can see and reject.
 */
const val DUPLICATE_BALANCE_WINDOW_MS = 48 * 60 * 60 * 1000L

private fun pair(a: Txn, b: Txn, kind: String, reason: String, auto: Boolean): LinkCandidate {
    val (x, y) = if (a.ref <= b.ref) a.ref to b.ref else b.ref to a.ref
    // The millisecond tie breaks on ref, so two devices deriving the same rows hide the same leg.
    val later = if (a.at > b.at || (a.at == b.at && a.ref > b.ref)) a else b
    return LinkCandidate(x, y, kind, reason, auto, later.ref)
}

/**
 * Duplicates, in this order, stopping at the first hit.
 *
 * The reference number is the bank's own idempotency key and settles it within its window. A
 * shared post-transaction balance settles it nearly as well: two genuinely separate identical
 * purchases cannot both leave the account holding the same figure — unless a month passes and
 * the same rent leaves the same account at the same figure again, which is why both matches
 * are time-bounded now. Anything weaker than that is never settled by the app, because two
 * identical taxi fares forty seconds apart is a thing that happens to people.
 */
fun findDuplicates(transactions: List<Txn>): List<LinkCandidate> {
    val ordered = transactions.sortedWith(compareBy({ it.at }, { it.ref }))

    val byRefNo = mutableListOf<LinkCandidate>()
    for (matches in ordered.filter { it.refNo.isNotEmpty() }.groupBy { it.bank to it.refNo }.values) {
        for (i in matches.indices) {
            for (j in i + 1 until matches.size) {
                if (matches[j].at - matches[i].at > DUPLICATE_REFNO_WINDOW_MS) break
                val a = matches[i]
                val b = matches[j]
                val compatible = a.accountId == b.accountId &&
                    a.direction != null && a.direction == b.direction &&
                    a.amountRial != null && a.amountRial == b.amountRial &&
                    a.signedRial != null && a.signedRial == b.signedRial &&
                    (a.mask.isBlank() || b.mask.isBlank() || a.mask == b.mask)
                byRefNo += pair(a, b, LinkKind.DUPLICATE, "refno", auto = compatible)
            }
        }
    }

    val byBalance = mutableListOf<LinkCandidate>()
    val balanceGroups = ordered
        .filter { it.signedRial != null && it.balanceRial != null }
        .groupBy { Triple(it.accountId, it.signedRial, it.balanceRial) }
    for (matches in balanceGroups.values) {
        for (i in matches.indices) {
            for (j in i + 1 until matches.size) {
                if (matches[j].at - matches[i].at > DUPLICATE_BALANCE_WINDOW_MS) break
                byBalance += pair(matches[i], matches[j], LinkKind.DUPLICATE, "balance", auto = true)
            }
        }
    }

    val nearby = mutableListOf<LinkCandidate>()
    for (i in ordered.indices) {
        val a = ordered[i]
        for (j in i + 1 until ordered.size) {
            val b = ordered[j]
            if (b.at - a.at > DUPLICATE_WINDOW_MS) break
            if (
                a.accountId == b.accountId &&
                a.signedRial != null &&
                a.signedRial == b.signedRial &&
                a.merchantNorm == b.merchantNorm
            ) {
                nearby += pair(a, b, LinkKind.DUPLICATE, "near", auto = false)
            }
        }
    }
    return (byRefNo + byBalance + nearby)
        .distinctBy { Triple(it.aRef, it.bRef, it.kind) }
}

/**
 * Transfers between two accounts she owns.
 *
 * Both legs must be in the ledger. A transfer to someone else's account is not a transfer, it
 * is an ordinary outflow, and treating it as one would make her spending disappear.
 *
 * Without this, moving fifty million between her own accounts reads as fifty million of income
 * and fifty million of spending — the same quietly-wrong total this app was built to avoid,
 * arriving from a new direction.
 */
private fun railOf(sent: Txn, received: Txn): String? = when {
    sent.channel in SLOW_RAILS -> sent.channel
    received.channel in SLOW_RAILS -> received.channel
    else -> null
}

private class TransferTimes(rows: List<Txn>) {
    private val ordered = rows.sortedBy { it.at }

    fun around(at: Long, window: Long): List<Txn> {
        val from = at.coerceAtLeast(Long.MIN_VALUE + window) - window
        val until = at.coerceAtMost(Long.MAX_VALUE - window) + window
        var low = 0
        var high = ordered.size
        while (low < high) {
            val middle = (low + high) ushr 1
            if (ordered[middle].at < from) low = middle + 1 else high = middle
        }
        val start = low
        high = ordered.size
        while (low < high) {
            val middle = (low + high) ushr 1
            if (ordered[middle].at <= until) low = middle + 1 else high = middle
        }
        return ordered.subList(start, low)
    }
}

fun findTransfers(transactions: List<Txn>): List<LinkCandidate> {
    val outgoing = transactions.filter { it.direction == "out" && it.amountRial != null }
    val incoming = transactions.filter { it.direction == "in" && it.amountRial != null }
    val allTimes = TransferTimes(incoming)
    val slowTimes = TransferTimes(incoming.filter { it.channel in SLOW_RAILS })
    val instantTimes = TransferTimes(incoming.filter { it.channel !in SLOW_RAILS })
    val candidates = outgoing.map { sent ->
        val nearby = if (sent.channel in SLOW_RAILS) {
            allTimes.around(sent.at, SLOW_RAIL_WINDOW_MS)
        } else {
            instantTimes.around(sent.at, TRANSFER_WINDOW_MS) +
                slowTimes.around(sent.at, SLOW_RAIL_WINDOW_MS)
        }
        sent to nearby.filter { received ->
            val slow = railOf(sent, received) != null
            received.accountId != sent.accountId &&
                kotlin.math.abs(received.at - sent.at) <=
                (if (slow) SLOW_RAIL_WINDOW_MS else TRANSFER_WINDOW_MS) &&
                if (slow) {
                    // پایا and ساتنا debit their fee separately, so the two legs are exactly
                    // equal or unrelated.
                    received.amountRial == sent.amountRial
                } else {
                    received.amountRial!! <= sent.amountRial!! &&
                        sent.amountRial - received.amountRial >= 0 &&
                        sent.amountRial - received.amountRial <= TRANSFER_FEE_MAX_RIAL
                }
        }
    }
    // One counter-leg settles at most one pair, so uniqueness is checked from the receiving
    // side too. Checked only from the sent side, two equal payments out inside the window of a
    // single equal leg in each looked unique from where they stood, both auto-paired with it,
    // and the one of them that had actually left the household vanished from every report.
    val claims = mutableMapOf<String, Int>()
    for ((_, matches) in candidates) {
        for (received in matches) claims[received.ref] = (claims[received.ref] ?: 0) + 1
    }
    val out = mutableListOf<LinkCandidate>()
    for ((sent, matches) in candidates) {
        if (matches.isEmpty()) continue
        // Ref breaks a tie in distance, so the pair does not depend on the order rows arrived.
        val nearest = matches.minWithOrNull(
            compareBy({ kotlin.math.abs(it.at - sent.at) }, { it.ref })
        )!!
        val rail = railOf(sent, nearest)
        val exact = nearest.amountRial == sent.amountRial
        val unique = matches.size == 1 && claims[nearest.ref] == 1
        // Automatic only when nothing about it is ambiguous: one candidate, claimed by no other
        // leg, exactly equal, and not on a rail whose forty-hour window makes coincidence
        // plausible.
        val auto = unique && exact && rail == null
        // The rail is its own reason, so the deck can say which one it noticed rather than
        // asserting a pairing she has no way to check.
        val reason = rail ?: if (exact) "exact" else "fee-band"
        out += pair(sent, nearest, LinkKind.TRANSFER, reason, auto)
    }
    return out
}

/**
 * Every candidate, with her verdicts applied last and winning outright.
 *
 * A rejection suppresses a pair for ever; a confirmation creates one the detectors missed.
 */
fun findLinks(
    transactions: List<Txn>,
    decisions: List<LinkDecision>,
): List<LinkCandidate> {
    val found = findDuplicates(transactions) + findTransfers(transactions)

    val rejected = decisions.filter { it.verdict == Verdict.REJECTED && !it.deleted }
        .map { Triple(it.aRef, it.bRef, it.kind) }.toSet()
    val confirmed = decisions.filter { it.verdict == Verdict.CONFIRMED && !it.deleted }

    val kept = found
        .distinctBy { Triple(it.aRef, it.bRef, it.kind) }
        .filterNot { Triple(it.aRef, it.bRef, it.kind) in rejected }

    // A confirmation may name legs the detectors never paired, so the later leg is looked up
    // rather than carried — and left blank when a leg is no longer in the ledger.
    val at = transactions.associate { it.ref to it.at }
    val hers = confirmed
        .filterNot { d -> kept.any { it.aRef == d.aRef && it.bRef == d.bRef && it.kind == d.kind } }
        .map { LinkCandidate(it.aRef, it.bRef, it.kind, "confirmed", auto = true, laterRef = laterOf(it.aRef, it.bRef, at)) }

    return kept.map { candidate ->
        // Her confirmation lifts a deck card into a settled fact.
        val settled = confirmed.any {
            it.aRef == candidate.aRef && it.bRef == candidate.bRef && it.kind == candidate.kind
        }
        if (settled) candidate.copy(auto = true, reason = "confirmed") else candidate
    } + hers
}

private fun laterOf(aRef: String, bRef: String, at: Map<String, Long>): String {
    val aAt = at[aRef] ?: return ""
    val bAt = at[bRef] ?: return ""
    return if (aAt > bAt || (aAt == bAt && aRef > bRef)) aRef else bRef
}

/** Both legs of every transfer that is settled, so reports can leave them out of both sides. */
fun transferRefs(links: List<LinkCandidate>): Set<String> = links
    .filter { it.kind == LinkKind.TRANSFER && it.auto }
    .flatMap { listOf(it.aRef, it.bRef) }
    .toSet()

/**
 * The later leg of every settled duplicate — hidden from reports, never deleted.
 *
 * [LinkCandidate.laterRef], not [LinkCandidate.bRef]: the pair is stored in ref order, a ref is
 * a hash, and hiding by hash order hid the original and kept the echo whenever the hashes
 * happened to sort against time. bRef remains the fallback for a confirmed pair whose legs the
 * ledger no longer holds, where "later" has no answer at all.
 */
fun hiddenRefs(links: List<LinkCandidate>): Set<String> = links
    .filter { it.kind == LinkKind.DUPLICATE && it.auto }
    .map { it.laterRef.ifEmpty { it.bRef } }
    .toSet()

@Dao
interface LinkCandidateDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun putAll(rows: List<LinkCandidate>)

    @Query("DELETE FROM link_candidate")
    suspend fun deleteAll()

    @Query("SELECT * FROM link_candidate")
    suspend fun all(): List<LinkCandidate>

    @Query("SELECT * FROM link_candidate WHERE a_ref = :ref OR b_ref = :ref")
    suspend fun touching(ref: String): List<LinkCandidate>
}

@Dao
interface LinkDecisionDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun put(row: LinkDecision)

    @Query("SELECT * FROM link_decision WHERE deleted = 0")
    suspend fun all(): List<LinkDecision>
}

File: pwa/src/links.ts
/**
 * Transfers and duplicates — the two ways one movement of money shows up as more than one
 * transaction, and the two ways a total goes quietly wrong if nothing notices (Links.kt).
 *
 * Every detector is a pure function of the transactions, and nothing guesses: a link counts only
 * when nothing about it is ambiguous — `auto` — and her verdict beats everything both ways. A near
 * miss is a non-auto candidate: it counts in nothing, and gives her verdict a pair to land on.
 */
import type { LinkDecision, Txn } from './model';

export const LinkKind = { TRANSFER: 'transfer', DUPLICATE: 'duplicate' } as const;
export const Verdict = { CONFIRMED: 'confirmed', REJECTED: 'rejected' } as const;

/** A pair the detectors propose, rebuilt on every derive. `aRef < bRef`. */
export interface LinkCandidate {
  aRef: string;
  bRef: string;
  kind: LinkDecision['kind'];
  /** Why, in a word, so the deck can say what it noticed rather than just assert it. */
  reason: string;
  /** True only when nothing about the match is ambiguous. */
  auto: boolean;
  /**
   * The leg that arrived second, which [hiddenRefs] strikes out — carried because refs are hashes
   * and their order against time is a coin flip. Blank only on a confirmation whose legs are gone.
   */
  laterRef: string;
}

/**
 * Her verdict on a pair. The id is `${kind}:${aRef}:${bRef}`, which is the phone's unique
 * (a_ref, b_ref, kind) index: without it `(a,b)` could be confirmed and rejected at once.
 */
export function linkDecision(aRef: string, bRef: string, kind: LinkDecision['kind'], verdict: LinkDecision['verdict'], now: number): LinkDecision {
  const [a, b] = aRef <= bRef ? [aRef, bRef] : [bRef, aRef];
  return { id: `${kind}:${a}:${b}`, aRef: a, bRef: b, kind, verdict, createdAt: now, updatedAt: now, deleted: false };
}

/** A fee band a bank may take out of a transfer without it stopping being the same money. */
export const TRANSFER_FEE_MAX_RIAL = 120_000;
/** Instant rails settle in seconds; anything past this is two separate movements. */
export const TRANSFER_WINDOW_MS = 15 * 60 * 1000;
/** پایا settles next business day and ساتنا in banking-hour cycles: too wide to ever trust silently. */
export const SLOW_RAIL_WINDOW_MS = 40 * 60 * 60 * 1000;
/** Read off either leg: the receiving bank writes «واریز پایا» while the sender says only «انتقال». */
const SLOW_RAILS = new Set(['paya', 'satna']);
/** Two identical purchases this close together are a real thing; only she may decide. */
export const DUPLICATE_WINDOW_MS = 90 * 1000;
/** Some banks re-print a contract number where a reference belongs: seven days, short of next month's instalment. */
export const DUPLICATE_REFNO_WINDOW_MS = 7 * 24 * 60 * 60 * 1000;
/** A shared post-transaction balance settles a duplicate only this close — or a monthly rent would hide for ever. */
export const DUPLICATE_BALANCE_WINDOW_MS = 48 * 60 * 60 * 1000;

const byTime = (a: Txn, b: Txn): number => a.at - b.at || (a.ref < b.ref ? -1 : a.ref > b.ref ? 1 : 0);

function pair(a: Txn, b: Txn, kind: LinkCandidate['kind'], reason: string, auto: boolean): LinkCandidate {
  const [x, y] = a.ref <= b.ref ? [a.ref, b.ref] : [b.ref, a.ref];
  // The millisecond tie breaks on ref, so two devices deriving the same rows hide the same leg.
  const later = a.at > b.at || (a.at === b.at && a.ref > b.ref) ? a : b;
  return { aRef: x, bRef: y, kind, reason, auto, laterRef: later.ref };
}

function groupBy<T>(items: readonly T[], key: (item: T) => string): T[][] {
  const groups = new Map<string, T[]>();
  for (const item of items) {
    const k = key(item);
    const group = groups.get(k);
    if (group) group.push(item); else groups.set(k, [item]);
  }
  return [...groups.values()];
}

const tripleOf = (c: { aRef: string; bRef: string; kind: string }): string => `${c.aRef}\u0000${c.bRef}\u0000${c.kind}`;

function distinct(candidates: LinkCandidate[]): LinkCandidate[] {
  const seen = new Set<string>();
  return candidates.filter((c) => !seen.has(tripleOf(c)) && seen.add(tripleOf(c)));
}

/**
 * Duplicates, in this order: the bank's reference number within its window, then a shared
 * post-transaction balance within its window, then — never settled by the app — the same money
 * at the same merchant ninety seconds apart, because two identical taxi fares happen to people.
 */
export function findDuplicates(transactions: readonly Txn[]): LinkCandidate[] {
  const ordered = [...transactions].sort(byTime);

  const byRefNo: LinkCandidate[] = [];
  for (const matches of groupBy(ordered.filter((t) => t.refNo !== ''), (t) => `${t.bank}\u0000${t.refNo}`)) {
    for (let i = 0; i < matches.length; i++) {
      for (let j = i + 1; j < matches.length; j++) {
        if (matches[j].at - matches[i].at > DUPLICATE_REFNO_WINDOW_MS) break;
        const a = matches[i];
        const b = matches[j];
        const compatible = a.accountId === b.accountId &&
          a.direction != null && a.direction === b.direction &&
          a.amountRial != null && a.amountRial === b.amountRial &&
          a.signedRial != null && a.signedRial === b.signedRial &&
          (!a.mask.trim() || !b.mask.trim() || a.mask === b.mask);
        byRefNo.push(pair(a, b, LinkKind.DUPLICATE, 'refno', compatible));
      }
    }
  }

  const byBalance: LinkCandidate[] = [];
  const balanceGroups = groupBy(
    ordered.filter((t) => t.signedRial != null && t.balanceRial != null),
    (t) => `${t.accountId}\u0000${t.signedRial}\u0000${t.balanceRial}`,
  );
  for (const matches of balanceGroups) {
    for (let i = 0; i < matches.length; i++) {
      for (let j = i + 1; j < matches.length; j++) {
        if (matches[j].at - matches[i].at > DUPLICATE_BALANCE_WINDOW_MS) break;
        byBalance.push(pair(matches[i], matches[j], LinkKind.DUPLICATE, 'balance', true));
      }
    }
  }

  const nearby: LinkCandidate[] = [];
  for (let i = 0; i < ordered.length; i++) {
    const a = ordered[i];
    for (let j = i + 1; j < ordered.length; j++) {
      const b = ordered[j];
      if (b.at - a.at > DUPLICATE_WINDOW_MS) break;
      if (a.accountId === b.accountId && a.signedRial != null && a.signedRial === b.signedRial && a.merchantNorm === b.merchantNorm) {
        nearby.push(pair(a, b, LinkKind.DUPLICATE, 'near', false));
      }
    }
  }
  return distinct([...byRefNo, ...byBalance, ...nearby]);
}

function railOf(sent: Txn, received: Txn): string | null {
  if (SLOW_RAILS.has(sent.channel)) return sent.channel;
  if (SLOW_RAILS.has(received.channel)) return received.channel;
  return null;
}

/** The rows within [window] of a moment, by binary search over a time-sorted copy. */
function timeIndex(rows: readonly Txn[]): (at: number, window: number) => Txn[] {
  const ordered = [...rows].sort((a, b) => a.at - b.at);
  const firstAt = (pred: (t: Txn) => boolean): number => {
    let low = 0;
    let high = ordered.length;
    while (low < high) {
      const middle = (low + high) >>> 1;
      if (pred(ordered[middle])) low = middle + 1; else high = middle;
    }
    return low;
  };
  return (at, window) => ordered.slice(firstAt((t) => t.at < at - window), firstAt((t) => t.at <= at + window));
}

/**
 * Transfers between two accounts she owns. Both legs must be in the ledger: money to someone
 * else's account is spending, and treating it as a transfer would make that spending disappear.
 */
export function findTransfers(transactions: readonly Txn[]): LinkCandidate[] {
  // A box move never leaves Blu, so it has no leg at another bank.
  const legs = transactions.filter((t) => t.channel !== 'box');
  const outgoing = legs.filter((t) => t.direction === 'out' && t.amountRial != null);
  const incoming = legs.filter((t) => t.direction === 'in' && t.amountRial != null);
  const allTimes = timeIndex(incoming);
  const slowTimes = timeIndex(incoming.filter((t) => SLOW_RAILS.has(t.channel)));
  const instantTimes = timeIndex(incoming.filter((t) => !SLOW_RAILS.has(t.channel)));
  const candidates = outgoing.map((sent) => {
    const nearby = SLOW_RAILS.has(sent.channel)
      ? allTimes(sent.at, SLOW_RAIL_WINDOW_MS)
      : [...instantTimes(sent.at, TRANSFER_WINDOW_MS), ...slowTimes(sent.at, SLOW_RAIL_WINDOW_MS)];
    const matches = nearby.filter((received) => {
      const slow = railOf(sent, received) != null;
      if (received.accountId === sent.accountId) return false;
      if (Math.abs(received.at - sent.at) > (slow ? SLOW_RAIL_WINDOW_MS : TRANSFER_WINDOW_MS)) return false;
      // پایا and ساتنا debit their fee separately, so the two legs are exactly equal or unrelated.
      if (slow) return received.amountRial === sent.amountRial;
      const fee = sent.amountRial! - received.amountRial!;
      return fee >= 0 && fee <= TRANSFER_FEE_MAX_RIAL;
    });
    return { sent, matches };
  });
  // One counter-leg settles at most one pair, so uniqueness is checked from the receiving side
  // too: two equal payments out cannot both claim the one leg that came in.
  const claims = new Map<string, number>();
  for (const { matches } of candidates) for (const r of matches) claims.set(r.ref, (claims.get(r.ref) ?? 0) + 1);
  const out: LinkCandidate[] = [];
  for (const { sent, matches } of candidates) {
    if (!matches.length) continue;
    // Ref breaks a tie in distance, so the pair does not depend on the order rows arrived.
    const nearest = matches.reduce((best, t) => {
      const d = Math.abs(t.at - sent.at) - Math.abs(best.at - sent.at);
      return d < 0 || (d === 0 && t.ref < best.ref) ? t : best;
    });
    const rail = railOf(sent, nearest);
    const exact = nearest.amountRial === sent.amountRial;
    const unique = matches.length === 1 && claims.get(nearest.ref) === 1;
    out.push(pair(sent, nearest, LinkKind.TRANSFER, rail ?? (exact ? 'exact' : 'fee-band'), unique && exact && rail == null));
  }
  return out;
}

/** Every candidate, with her verdicts applied last: a rejection suppresses a pair for ever, a confirmation creates one. */
export function findLinks(transactions: readonly Txn[], decisions: readonly LinkDecision[]): LinkCandidate[] {
  const found = [...findDuplicates(transactions), ...findTransfers(transactions)];
  const live = decisions.filter((d) => !d.deleted);
  const rejected = new Set(live.filter((d) => d.verdict === Verdict.REJECTED).map(tripleOf));
  const confirmed = new Set(live.filter((d) => d.verdict === Verdict.CONFIRMED).map(tripleOf));
  const kept = distinct(found).filter((c) => !rejected.has(tripleOf(c)));
  const keptTriples = new Set(kept.map(tripleOf));

  // A confirmation may name legs the detectors never paired, so its later leg is looked up.
  const at = new Map(transactions.map((t) => [t.ref, t.at]));
  const hers = live
    .filter((d) => d.verdict === Verdict.CONFIRMED && !keptTriples.has(tripleOf(d)))
    .map((d): LinkCandidate => ({ aRef: d.aRef, bRef: d.bRef, kind: d.kind, reason: 'confirmed', auto: true, laterRef: laterOf(d.aRef, d.bRef, at) }));

  return [
    ...kept.map((c) => (confirmed.has(tripleOf(c)) ? { ...c, auto: true, reason: 'confirmed' } : c)),
    ...hers,
  ];
}

function laterOf(aRef: string, bRef: string, at: ReadonlyMap<string, number>): string {
  const aAt = at.get(aRef);
  const bAt = at.get(bRef);
  if (aAt == null || bAt == null) return '';
  return aAt > bAt || (aAt === bAt && aRef > bRef) ? aRef : bRef;
}

/** Both legs of every settled transfer, so reports leave them out of both sides. */
export function transferRefs(links: readonly LinkCandidate[]): Set<string> {
  return new Set(links.filter((l) => l.kind === LinkKind.TRANSFER && l.auto).flatMap((l) => [l.aRef, l.bRef]));
}

/** The later leg of every settled duplicate — hidden from reports, never deleted. */
export function hiddenRefs(links: readonly LinkCandidate[]): Set<string> {
  return new Set(links.filter((l) => l.kind === LinkKind.DUPLICATE && l.auto).map((l) => l.laterRef || l.bRef));
}
